openedx/core/djangoapps/user_api/preferences/api.py

```diff
@@ -0,0 +1,333 @@
+        username (str): Optional username for which to look up the preferences. If not specified,
+            `requesting_user.username` is assumed.
+
+    Returns:
+         A dict containing account fields.
+
+    Raises:
+         UserNotFound: no user with username `username` exists (or `requesting_user.username` if
+            `username` is not specified)
+         UserNotAuthorized: the requesting_user does not have access to the user preference.
+         UserAPIInternalError: the operation failed due to an unexpected error.
+    """
+    existing_user = _get_user(requesting_user, username, allow_staff=True)
+    user_serializer = UserSerializer(existing_user)
+    return user_serializer.data["preferences"]
+
+
+@intercept_errors(UserAPIInternalError, ignore_errors=[UserAPIRequestError])
+@transaction.commit_on_success
+def update_user_preferences(requesting_user, update, username=None):
+    """Update the user preferences for the given username.
+
+    Note:
+        It is up to the caller of this method to enforce the contract that this method is only called
+        with the user who made the request.
+
+    Arguments:
+        requesting_user (User): The user requesting to modify account information. Only the user with username
+            'username' has permissions to modify account information.
+        update (dict): The updated account field values.
+            Some notes:
+                Values are expected to be strings. Non-string values will be converted to strings.
+                Null values for a preference will be treated as a request to delete the key in question.
+        username (string): Optional username specifying which account should be updated. If not specified,
+            `requesting_user.username` is assumed.
+
+    Raises:
+        UserNotFound: no user with username `username` exists (or `requesting_user.username` if
+            `username` is not specified)
+        UserNotAuthorized: the requesting_user does not have access to change the account
+            associated with `username`
+        PreferenceValidationError: the update was not attempted because validation errors were found
+        PreferenceUpdateError: the operation failed when performing the update.
+        UserAPIInternalError: the operation failed due to an unexpected error.
+    """
+    existing_user = _get_user(requesting_user, username)
+
+    # First validate each preference setting
+    errors = {}
+    for preference_key in update.keys():
+        preference_value = update[preference_key]
+        if preference_value is not None:
+            try:
+                _validate_user_preference(existing_user, preference_key, preference_key)
+            except PreferenceValidationError as error:
+                developer_message = error.preference_errors[preference_key]["developer_message"]
+                errors[preference_key] = {
+                    "developer_message": developer_message
```

Comment:
I decided I didn't have a localized user message so I wasn't returning one. Right now we aren't really validating the values, and user's can provide their own keys, so I don't know if this matters. However, I suppose it would be good to normalize this result.